compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+                if (checkFile.exists) List(checkFile)
+                else Nil
+              }
+            }.headOption
+            val reporter = compile(sourceFiles, flags, false, outDir)
+
+            if (reporter.errorCount > 0) {
+              fail()
+              val errorMsg = target.buildInstructions(reporter.errorCount, reporter.warningCount)
+              addFailureInstruction(errorMsg)
+              failTarget(target)
+              reporter.echo(errorMsg)
+              reporter.flushToFile()
+            }
+
+            completeCompilation(reporter.errorCount)
+            (reporter.errorCount, reporter.warningCount, checkFile.isDefined, () => verifyOutput(checkFile.get, outDir, target, reporter.warningCount))
+          }
+
+          case target @ SeparateCompilationTarget(dir, flags, outDir) => {
+            val checkFile = new JFile(dir.getAbsolutePath.reverse.dropWhile(_ == '/').reverse + ".check")
+            val (errorCount, warningCount) =
+              target
+                .compilationUnits
+                .map(files => compile(files.filter(isCompilable), flags, false, outDir))
+                .foldLeft((0,0)) { case ((errors, warnings), reporter) =>
+                  if (reporter.errorCount > 0) {
+                    val errorMsg = target.buildInstructions(reporter.errorCount, reporter.warningCount)
+                    addFailureInstruction(errorMsg)
+                    reporter.echo(errorMsg)
+                    reporter.flushToFile()
+                  }
+
+                  (errors + reporter.errorCount, warnings + reporter.warningCount)
+                }
+
+            if (errorCount > 0) fail()
+
+            completeCompilation(errorCount)
+            (errorCount, warningCount, checkFile.exists, () => verifyOutput(checkFile, outDir, target, warningCount))
+          }
+        }
+
+        if (errorCount == 0 && hasCheckFile) doVerify()
+        else if (errorCount > 0) {
+          echo(s"\nCompilation failed for: '$target'")
+          val buildInstr = target.buildInstructions(errorCount, warningCount)
+          addFailureInstruction(buildInstr)
+          failTarget(target)
+        }
+      }
+    }
+  }
+
+  private final class NegTest(targets: List[Target], times: Int, threadLimit: Option[Int], suppressAllOutput: Boolean)
+  extends Test(targets, times, threadLimit, suppressAllOutput) {
+    protected def compilationRunnable(target: Target): Runnable = new Runnable {
+      def run(): Unit = compileTry {
+        // In neg-tests we allow two types of error annotations,
+        // "nopos-error" which doesn't care about position and "error" which
+        // has to be annotated on the correct line number.
+        //
+        // We collect these in a map `"file:row" -> numberOfErrors`, for
+        // nopos errors we save them in `"file" -> numberOfNoPosErrors`
+        def errorMapAndExpected(files: Array[JFile]): (HashMap[String, Integer], Int) = {
```

Comment:
collectErrorMapAndExpectedCount.
Also, using strings as keys in map isn't the best idea. VM isn't prohibited from re-using strings unless you explicitly call `new`.